### PR TITLE
Obtain a remote file existence, mtime, and size inventory instead of querying this information file by file.

### PR DESCRIPTION
**Pull request by johanneskoester**
@vsoch, I think everything is prepared such that you just need to put your code where I added the TODO.

### CODE REVIEW

**Comment by vsoch** [comment]
Aye aye!

**Comment by vsoch** [comment]
A quick question - I have an implementation that works fairly well in the case that the storage is small (e.g., snakemake-testing) but when we go to retrieve a file in public storage (e.g., one of the prefixed with `gcp-public-data-landsat/LC08/01/001` the method we chose is going to iterate over the entire bucket which is HUGE and it turns a rather small / trivial run to something that could take hours I think. So - how should we address this? I tried added a check for prefix, but that still requires listing all the blobs. I'll push what I have so far shortly and we can discuss.

**Comment by vsoch** [comment]
oup nevermind figured it out, I can add a prefix :)

**Comment by vsoch** [comment]
okay here is a first shot - not a ton of code but seems to generate the structure as specified. Please take a look, and happy Friday! https://github.com/snakemake/snakemake/pull/423/commits/0a069ca554d36dc752dbae76642723dabe842386

File: snakemake/remote/GS.py [hunk 1]
```diff
@@ -125,6 +126,27 @@ def __init__(
...
+            - cache.exist_remote
+            - cache_mtime
+            - cache.size
+        """
+        for blob in self.client.list_blobs(
+            self.bucket_name, prefix=os.path.dirname(self.blob.name)
+        ):
+
+            # By way of being listed, it exists. mtime is a datetime object
+            name = "%s/%s" % (blob.bucket.name, blob.name)
```
Comment (johanneskoester) [inline snakemake/remote/GS.py:145]: Please use the .format method instead, old python string formatting might go away at some point.

Comment (vsoch) [reply snakemake/remote/GS.py]: All set.

**Comment by vsoch** [comment]
@johanneskoester for the benchmark - two quick questions.

 - are you referring to the dna-seq-varlociraptor,  kim-wxs-varlociraptor, or something else?
 - for the benchmark - do you want something added as a test or just run locally / on a cloud instance?

**Comment by vsoch** [comment]
AWS execution failed (I think unrelated to this PR) 
image

**Comment by johanneskoester** [comment]
> @johanneskoester for the benchmark - two quick questions.
> 
>     * are you referring to the dna-seq-varlociraptor,  kim-wxs-varlociraptor, or something else?
> 
>     * for the benchmark - do you want something added as a test or just run locally / on a cloud instance?

kim-wxs-varlociraptor. You should probably add time measurement to the relevant part, namely here: https://github.com/snakemake/snakemake/blob/master/snakemake/workflow.py#L655

I want to know if the dag.init() method is faster now. Whether you run this locally or on gcloud does not matter, as long as you use --default-remote-provider GS and --default-remote-prefix set to the corresponding bucket where you stored the previous results.

**Comment by vsoch** [comment]
Oh that's even easier than what I have in mind! Note that I ran it three times in case anything with the google cloud API needed warm up (and I think this is likely because the first run was 8 seconds, then the next 2 were 7, and then the "recorded" three were in the 6s. Here is three runs with the inventory branch:
```
Building DAG of jobs...
Total elapsed seconds: 6.951563358306885
Total elapsed seconds: 6.76206636428833
Total elapsed seconds: 6.492744207382202
```
and here is the current master branch

```
Total elapsed seconds: 3.7220818996429443
Total elapsed seconds: 3.218263626098633
Total elapsed seconds: 3.3768815994262695
```
So it's actually slower, and I think this is fairly logical to deduce why - for the inventory version, instead of just doing blob.exists for a smaller set of specific items, we are doing it for every file in some prefix of the bucket where we reference files. I think we would only see a speedup in the case that we needed to actually do blob.exists for a much larger set of files. Is there a larger dataset we can use to test this instead? If that is true, I think we'd also want it to be an optional flag (and not the default) because most analyses probably don't have that many files.

**Comment by vsoch** [comment]
Should we perhaps only cache the files that are needed based on blob.exists, and not the entire prefix?

**Comment by vsoch** [comment]
Inventory without going through the bucket prefix with list_blobs (about the same if not slower) on the inventory branch.
```
Total elapsed seconds: 3.5445785522460938
Total elapsed seconds: 3.7110350131988525
Total elapsed seconds: 3.5723507404327393
```
Going back to sleep now but of course will check in after I wake up!

**Comment by johanneskoester** [comment]
How many samples do you have in the kim-wxs workflow?

**Comment by vsoch** [comment]
I believe it’s still at this state (N=2) https://github.com/koesterlab/kim-wxs-varlociraptor/blob/master/config/samples.tsv.

**Comment by vsoch** [comment]
okay, I'm not totally sure how it's setup to get data, but I found the original data file in the paper associated with the repo, and tried to produce another samples.tsv that includes all the other identifiers:

```
sample_name	alias	group	platform	purity
C1	tumor	1	ILLUMINA	1
N1	normal	1	ILLUMINA	
C100	tumor	100	ILLUMINA	1
N100	normal	100	ILLUMINA	
C104	tumor	104	ILLUMINA	1
N104	normal	104	ILLUMINA	
C105	tumor	105	ILLUMINA	1
N105	normal	105	ILLUMINA	
C106	tumor	106	ILLUMINA	1
N106	normal	106	ILLUMINA	
C107	tumor	107	ILLUMINA	1
N107	normal	107	ILLUMINA	
C108	tumor	108	ILLUMINA	1
N108	normal	108	ILLUMINA	
C109	tumor	109	ILLUMINA	1
N109	normal	109	ILLUMINA	
C111	tumor	111	ILLUMINA	1
N111	normal	111	ILLUMINA	
C113	tumor	113	ILLUMINA	1
N113	normal	113	ILLUMINA	
C114	tumor	114	ILLUMINA	1
N114	normal	114	ILLUMINA	
C115	tumor	115	ILLUMINA	1
N115	normal	115	ILLUMINA	
C116	tumor	116	ILLUMINA	1
N116	normal	116	ILLUMINA	
C118	tumor	118	ILLUMINA	1
N118	normal	118	ILLUMINA	
C119	tumor	12	ILLUMINA	1
N119	normal	12	ILLUMINA	
C12	tumor	120	ILLUMINA	1
N12	normal	120	ILLUMINA	
C120	tumor	120	ILLUMINA	1
N20	normal	120	ILLUMINA	
C121	tumor	121	ILLUMINA	1
N121	normal	121	ILLUMINA	
C122	tumor	122	ILLUMINA	1
N122	normal	122	ILLUMINA	
C124	tumor	124	ILLUMINA	1
N124	normal	124	ILLUMINA	
C125	tumor	125	ILLUMINA	1
N125	normal	125	ILLUMINA	
C129	tumor	129	ILLUMINA	1
N129	normal	129	ILLUMINA	
C130	tumor	130	ILLUMINA	1
N130	normal	130	ILLUMINA	
C14	tumor	14	ILLUMINA	1
N14	normal	14	ILLUMINA	
C15	tumor	15	ILLUMINA	1
N15	normal	15	ILLUMINA	
C16	tumor	16	ILLUMINA	1
N16	normal	16	ILLUMINA	
C17	tumor	17	ILLUMINA	1
N17	normal	17	ILLUMINA	
C18	tumor	18	ILLUMINA	1
N18	normal	18	ILLUMINA	
C19	tumor	19	ILLUMINA	1
N19	normal	19	ILLUMINA	
C2	tumor	2	ILLUMINA	1
N2	normal	2	ILLUMINA	
C20	tumor	20	ILLUMINA	1
N20	normal	20	ILLUMINA	
C21	tumor	21	ILLUMINA	1
N21	normal	21	ILLUMINA	
C23	tumor	23	ILLUMINA	1
N23	normal	23	ILLUMINA	
C24	tumor	24	ILLUMINA	1
N24	normal	24	ILLUMINA	
C25	tumor	25	ILLUMINA	1
N25	normal	25	ILLUMINA	
C26	tumor	26	ILLUMINA	1
N26	normal	26	ILLUMINA	
C27	tumor	27	ILLUMINA	1
N27	normal	27	ILLUMINA	
C28	tumor	28	ILLUMINA	1
N28	normal	28	ILLUMINA	
C29	tumor	29	ILLUMINA	1
N29	normal	29	ILLUMINA	
C3	tumor	3	ILLUMINA	1
N3	normal	3	ILLUMINA	
C30	tumor	30	ILLUMINA	1
N30	normal	30	ILLUMINA	
C31	tumor	31	ILLUMINA	1
N31	normal	31	ILLUMINA	
C32	tumor	32	ILLUMINA	1
N32	normal	32	ILLUMINA	
C34	tumor	34	ILLUMINA	1
N34	normal	34	ILLUMINA	
C36	tumor	36	ILLUMINA	1
N36	normal	36	ILLUMINA	
C37	tumor	37	ILLUMINA	1
N37	normal	37	ILLUMINA	
C38	tumor	38	ILLUMINA	1
N38	normal	38	ILLUMINA	
C39	tumor	39	ILLUMINA	1
N39	normal	39	ILLUMINA	
C4	tumor	4	ILLUMINA	1
N4	normal	4	ILLUMINA	
C40	tumor	40	ILLUMINA	1
N40	normal	40	ILLUMINA	
C42	tumor	42	ILLUMINA	1
N42	normal	42	ILLUMINA	
C43	tumor	43	ILLUMINA	1
N43	normal	43	ILLUMINA	
C44	tumor	44	ILLUMINA	1
N44	normal	44	ILLUMINA	
C45	tumor	45	ILLUMINA	1
N45	normal	45	ILLUMINA	
C46	tumor	46	ILLUMINA	1
N46	normal	46	ILLUMINA	
C47	tumor	47	ILLUMINA	1
N47	normal	47	ILLUMINA	
C48	tumor	48	ILLUMINA	1
N48	normal	48	ILLUMINA	
C49	tumor	49	ILLUMINA	1
N49	normal	49	ILLUMINA	
C5	tumor	5	ILLUMINA	1
N5	normal	5	ILLUMINA	
C50	tumor	50	ILLUMINA	1
N50	normal	50	ILLUMINA	
C53	tumor	53	ILLUMINA	1
N53	normal	53	ILLUMINA	
C54	tumor	54	ILLUMINA	1
N54	normal	54	ILLUMINA	
C55	tumor	55	ILLUMINA	1
N55	normal	55	ILLUMINA	
C57	tumor	57	ILLUMINA	1
N57	normal	57	ILLUMINA	
C58	tumor	58	ILLUMINA	1
N58	normal	58	ILLUMINA	
C59	tumor	59	ILLUMINA	1
N59	normal	59	ILLUMINA	
C6	tumor	6	ILLUMINA	1
N6	normal	6	ILLUMINA	
C60	tumor	60	ILLUMINA	1
N60	normal	60	ILLUMINA	
C61	tumor	61	ILLUMINA	1
N61	normal	61	ILLUMINA	
C62	tumor	62	ILLUMINA	1
N62	normal	62	ILLUMINA	
C63	tumor	63	ILLUMINA	1
N63	normal	63	ILLUMINA	
C64	tumor	64	ILLUMINA	1
N64	normal	64	ILLUMINA	
C65	tumor	65	ILLUMINA	1
N65	normal	65	ILLUMINA	
C66	tumor	66	ILLUMINA	1
N66	normal	66	ILLUMINA	
C67	tumor	67	ILLUMINA	1
N67	normal	67	ILLUMINA	
C68	tumor	68	ILLUMINA	1
N68	normal	68	ILLUMINA	
C7	tumor	7	ILLUMINA	1
N7	normal	7	ILLUMINA	
C70	tumor	70	ILLUMINA	1
N70	normal	70	ILLUMINA	
C72	tumor	72	ILLUMINA	1
N72	normal	72	ILLUMINA	
C73	tumor	73	ILLUMINA	1
N73	normal	73	ILLUMINA	
C75	tumor	75	ILLUMINA	1
N75	normal	75	ILLUMINA	
C78	tumor	78	ILLUMINA	1
N78	normal	78	ILLUMINA	
C8	tumor	8	ILLUMINA	1
N8	normal	8	ILLUMINA	
C80	tumor	80	ILLUMINA	1
N80	normal	80	ILLUMINA	
C81	tumor	81	ILLUMINA	1
N81	normal	81	ILLUMINA	
C82	tumor	82	ILLUMINA	1
N82	normal	82	ILLUMINA	
C83	tumor	83	ILLUMINA	1
N83	normal	83	ILLUMINA	
C84	tumor	84	ILLUMINA	1
N84	normal	84	ILLUMINA	
C85	tumor	85	ILLUMINA	1
N85	normal	85	ILLUMINA	
C86	tumor	86	ILLUMINA	1
N86	normal	86	ILLUMINA	
C87	tumor	87	ILLUMINA	1
N87	normal	87	ILLUMINA	
C89	tumor	89	ILLUMINA	1
N89	normal	89	ILLUMINA	
C90	tumor	90	ILLUMINA	1
N90	normal	90	ILLUMINA	
C92	tumor	92	ILLUMINA	1
N92	normal	92	ILLUMINA	
C93	tumor	93	ILLUMINA	1
N93	normal	93	ILLUMINA	
C94	tumor	94	ILLUMINA	1
N94	normal	94	ILLUMINA	
C95	tumor	95	ILLUMINA	1
N95	normal	95	ILLUMINA	
C96	tumor	96	ILLUMINA	1
N96	normal	96	ILLUMINA	
C97	tumor	97	ILLUMINA	1
N97	normal	97	ILLUMINA	
C98	tumor	98	ILLUMINA	1
N98	normal	98	ILLUMINA	
C99	tumor	99	ILLUMINA	1
N99	normal	99	ILLUMINA	
```
There is an issue with get_read_group:
```
Building DAG of jobs...
InputFunctionException in line 1 of /home/vanessa/snakemake-work/kim-wxs-varlociraptor/workflow/rules/mapping.smk:
KeyError: 'C1'
Wildcards:
sample=C1
```
and I think it's because I need to also edit units.tsv
```
sample_name	unit_name	fq1	fq2	adapters	sra
C23	C23			-a CTGTCTCTTGATCACA	SRR5882466
N23	N23			-a CTGTCTCTTGATCACA	SRR5882453
```
I'm pretty sure those SRR numbers correspond to download identifiers - I don't remember where they come from! I think if we can generate this units.tsv file we should be able to run this more at scale?

**Comment by vsoch** [comment]
okay I was able to hack together a fake lookup for units.tsv, and here are some quick times for the same situations above!

```
# using the inventory, as currently implemented with list_blobs:
Building DAG of jobs...
Total elapsed seconds: 488.43199157714844

# the current master branch (without inventory)
Building DAG of jobs...
Total elapsed seconds: 225.8770215511322

# Adding each individually to the cache via blob.exists()
Building DAG of jobs...
Total elapsed seconds: 229.28575086593628
```

I'm thinking the implementation to cache the entire prefix of a bucket is off, and instead we should be doing the third strategy of caching _a specific remote file only when we need it_ which means we won't use the list_blobs function as we previously discussed. Attached here are my hacky files - @johanneskoester let me know how you would like to proceed.

[units.tsv.txt](https://github.com/snakemake/snakemake/files/4725779/units.tsv.txt)
[samples.tsv.txt](https://github.com/snakemake/snakemake/files/4725780/samples.tsv.txt)

**Comment by johanneskoester** [comment]
Before we decide any further, we must understand where list_blobs spends so much time. Is it inside of the google API, inside of the method of the snakemake remote provider? Or maybe list_blobs actually does multiple api calls internally? One option to find this out is to use a python profiler. Snakemake has a builtin profile mechanism, that can be activated with `--runtime-profile FILE`. Please run that, in dryrun, and post the results here.

**Comment by vsoch** [comment]
okay, so I needed to install a library called yappi and then do a dryrun with the profile command for the inventory branch:

```bash
$ pip install yappi
$ snakemake --google-lifesciences --default-remote-prefix snakemake-testing/kims-v-inventory --use-conda --google-lifesciences-keep-cache --google-lifesciences-region us-west1 --dryrun --runtime-profile inventory-runtime-profile.txt
```
Note that this is with the really big list of subjects that I hacked together, and not the original small one with N=2. Here you go:

[inventory-runtime-profile.txt](https://github.com/snakemake/snakemake/files/4731428/inventory-runtime-profile.txt)

**Comment by vsoch** [comment]
I found the list blobs function [here](https://github.com/googleapis/python-storage/blob/b26f9fa8a767b7d5affea8d2c4b87163ce979fd2/google/cloud/storage/bucket.py#L1050) (note that it used to be with the google-python-client module and was moved. That uses a google.api_core.page_iterator.Iterator, [here](https://github.com/googleapis/python-api-core/blob/master/google/api_core/page_iterator.py#L151) which mostly seems to help with pagination of standard Blob objects.

**Comment by johanneskoester** [comment]
Something is odd here. Client.list_blobs was called 14 thousand times...

**Comment by johanneskoester** [comment]
I think I know what is going on. Let me think a bit.

**Comment by johanneskoester** [comment]
@vsoch ok, I think I might have fixed the issue. The cache now knows whether an inventory has been created for the root (the bucket) of a file. Hence, we can check whether for any other file in the blob a new inventory has to be created.

File: snakemake/io.py [hunk 2]
```diff
@@ -83,13 +83,28 @@ def __init__(self):
...
+        # Indicator whether an inventory has been created for the root of a given IOFile.
+        # In case of remote objects the root is the bucket or server host.
+        self.has_inventory = set()
         self.active = True
 
+    @classmethod
+    def get_inventory_root(cls, path):
+        """If eligible for inventory, get the root of a given path."""
+        path = Path(path)
+        if not path.is_absolute(path) and path.parts[0] != "..":
```
Comment (vsoch) [inline snakemake/io.py]:
```suggestion
        if not path.is_absolute() and path.parts[0] != "..":
```

**Comment by vsoch** [comment]
okay, new runtime profile is attached (for inventory), took about a minute:
```
Building DAG of jobs...
Total time is 66.83892107009888 seconds
```
[inventory-runtime-remote-prefix-updated.txt](https://github.com/snakemake/snakemake/files/4744726/inventory-runtime-remote-prefix-updated.txt)

And the current master (without inventory) is slower!

```
Building DAG of jobs...
Total time is 314.06817507743835 seconds
```
Hopefully that helps, need to go back to bed for now :)

**Comment by johanneskoester** [comment]
Awesome, that's a very nice speedup!

**Comment by johanneskoester** [comment]
Can you benchmark again? I took the opportunity to add some more performance optimizations.

**Comment by johanneskoester** [comment]
And another optimization.

**Comment by johanneskoester** [comment]
And yet another optimization.

**Comment by vsoch** [comment]
Should there be a significant difference in using the benchmark or not? I did a run first with the benchmark - it was slightly slower:
```
Building DAG of jobs...
Total time is 94.54841041564941 seconds
```
[inventory-runtime-profile-optimized.txt](https://github.com/snakemake/snakemake/files/4753172/inventory-runtime-profile-optimized.txt)

and then I removed it and ran two times, it was hugely speedy:
```
Building DAG of jobs...
Total time is 13.880852937698364 seconds

Building DAG of jobs...
Total time is 13.960675954818726 seconds
```
Thinking that some "warming up" must be going on, I added it back to record a metrics file for the faster cases, but that actually slowed it down again :p 

```
Building DAG of jobs...
Total time is 96.95001125335693 seconds
```
Just to make sure it wasn't chance, ran it two more times without the profile:

```
Building DAG of jobs...
Total time is 13.745625257492065 seconds

Building DAG of jobs...
Total time is 13.872326374053955 seconds
```
So I suspect I can't get a profile for the faster case. :) Is this expected?

**Comment by johanneskoester** [comment]
The profiling adds a lot of runtime overhead, hence it will always be slower than without profiling. But the comparison with the previous benchmark surprises me. In particular when looking again at https://github.com/snakemake/snakemake/pull/423#issuecomment-640452061, I noticed that the total time (112s) of the root function does not correspond to the runtime that you mentioned in your post (66s). This does not seem to be the case for the latest benchmark (94s vs 97s). Just to be sure there is not mixup, can you repeat the benchmark for commit 86c1f37?

**Comment by vsoch** [comment]
Sure thing.

```
$ snakemake --version
5.19.0+10.g86c1f37c.dirty
```
running:

```
$ snakemake --google-lifesciences --default-remote-prefix snakemake-testing/kims-v-inventory --use-conda --google-lifesciences-keep-cache --google-lifesciences-region us-west1 --dryrun --runtime-profile inventory-runtime-profile-commit-86c.txt
Building DAG of jobs...
Total time is 68.69542264938354 seconds
```
[inventory-runtime-profile-commit-86c.txt](https://github.com/snakemake/snakemake/files/4757681/inventory-runtime-profile-commit-86c.txt)

**Comment by johanneskoester** [comment]
Great, thanks. Now it is consistent again. The slowdown comes from the fact that my optimization has introduced a cache for output modification times. This, in turn, has led to a lot of evaluations of `Job.__eq__`, which was comparably slow. I have therefore added another optimization. As that is quite a major change in how Snakemake handles jobs, I've put it into a separate branch (PR #438).

Bottom line, could you benchmark the branch fast-job-hash`?

**Comment by vsoch** [comment]
That indeed is faster!
```
Building DAG of jobs...
Total time is 31.17263126373291 seconds
```
[inventory-runtime-fast-job-hash.txt](https://github.com/snakemake/snakemake/files/4764348/inventory-runtime-fast-job-hash.txt)

**Comment by johanneskoester** [comment]
> That indeed is faster!
> 
> ```
> Building DAG of jobs...
> Total time is 31.17263126373291 seconds
> ```
> 
> [inventory-runtime-fast-job-hash.txt](https://github.com/snakemake/snakemake/files/4764348/inventory-runtime-fast-job-hash.txt)

So, 7 times faster than the master branch. Great!

For comparing, can you benchmark a dry-run on the same machine without --default-remote-provider? I might quickly implement inventory for local files as well then.

**Comment by vsoch** [comment]
Yep! Do you want me to use the inventory branch, or the fast-job-hash (or have you merged fast-job-hash into here?).

**Comment by johanneskoester** [comment]
> Yep! Do you want me to use the inventory branch, or the fast-job-hash (or have you merged fast-job-hash into here?).

You can use this branch, I have already merged fast-job-hash into it.

**Comment by vsoch** [comment]
okay, so first is the run with `--default-remote-prefix` but the inventory branch updated:

```
$ snakemake --google-lifesciences --default-remote-prefix snakemake-testing/kims-v-inventory --use-conda --google-lifesciences-keep-cache --google-lifesciences-region us-west1 --dryrun --runtime-profile inventory-runtime-fast-job-inventory.txt
Building DAG of jobs...
Total time is 30.661144018173218 seconds
```
[inventory-runtime-fast-job-inventory.txt](https://github.com/snakemake/snakemake/files/4767651/inventory-runtime-fast-job-inventory.txt)

Next is the same branch without the prefix (seems about the same):
```
$  snakemake  --use-conda --dryrun --runtime-profile inventory-runtime-fast-job-inventory-local.txt
Building DAG of jobs...
Total time is 29.053742170333862 seconds
```
[inventory-runtime-fast-job-inventory-local.txt](https://github.com/snakemake/snakemake/files/4767681/inventory-runtime-fast-job-inventory-local.txt)

**Comment by johanneskoester** [comment]
@vsoch thanks a lot. Can you now benchmark locally again, with the latest commit (which includes an inventory implementation for local files)? The benefit should be small on your setup (you obviously have a very fast FS), but I expect it to be quite large on network file systems. Anyway, I just want to be sure that there is an effect even on your system.

**Comment by vsoch** [comment]
It seems a hair bit slower, I did this three times to be sure (profile is from the last)
```
$ snakemake  --use-conda --dryrun --runtime-profile inventory-runtime-local.txtBuilding DAG of jobs...
Total time is 33.870110750198364 seconds

Building DAG of jobs...
Total time is 33.37542676925659 seconds

Building DAG of jobs...
Total time is 33.51922798156738 seconds
```
[inventory-runtime-local.txt](https://github.com/snakemake/snakemake/files/4770295/inventory-runtime-local.txt)

and I ran it three times without a profiler:

```
$ snakemake  --use-conda --dryrun 
Building DAG of jobs...
Total time is 10.887972116470337 seconds

Building DAG of jobs...
Total time is 10.832780361175537 seconds

Building DAG of jobs...
Total time is 10.585854291915894 seconds
```
Which is a tad slower than current master

```'
$ snakemake  --use-conda --dryrun 
Building DAG of jobs...
Total time is 7.906724691390991 seconds

Building DAG of jobs...
Total time is 7.6367716789245605 seconds

Building DAG of jobs...
Total time is 7.569443225860596 seconds
```

I was curious so I ran it on one of our cluster nodes, but probably this should be disregarded because doing anything on there is so slow... just to install the branch I had to leave and come back it was taking so long!

```
$ /home/users/vsochat/.local/bin/snakemake --use-conda --dryrun
Building DAG of jobs...
Total time: 19.73143768310547

Building DAG of jobs...
Total time: 10.78032660484314

Building DAG of jobs...
Total time: 11.043591260910034
```
and with profiler
```
$ /home/users/vsochat/.local/bin/snakemake --use-conda --dryrun --runtime-profile cluster-profile.txt
Building DAG of jobs...
Total time: 24.96830153465271
```
Ug, it's always so frustrating to use our resources... I really feel for our users!
[cluster-profile.txt](https://github.com/snakemake/snakemake/files/4770426/cluster-profile.txt)

**Comment by johanneskoester** [comment]
@vsoch can you benchmark locally and remote again? I have removed another superfluous call.

**Comment by vsoch** [comment]
Sure thing! Here is local first
```
$ snakemake  --use-conda --dryrun --runtime-profile inventory-runtime-local-june-15.txt
Building DAG of jobs...
Total time is: 26.46983242034912 seconds.
```
[inventory-runtime-local-june-15.txt](https://github.com/snakemake/snakemake/files/4781394/inventory-runtime-local-june-15.txt)

and remote 
```bash
$ snakemake --google-lifesciences --default-remote-prefix snakemake-testing/kims-v-inventory --use-conda --google-lifesciences-keep-cache --google-lifesciences-region us-west1 --dryrun --runtime-profile inventory-runtime-gls-june-15.txt
Building DAG of jobs...
Total time is: 29.808923959732056 seconds.
```

[inventory-runtime-gls-june-15.txt](https://github.com/snakemake/snakemake/files/4781399/inventory-runtime-gls-june-15.txt)

**Comment by johanneskoester** [comment]
and local master with --runtime-profile?

**Comment by vsoch** [comment]
Sure! Let me know all at once next time since I have to start up and shut down an instance.
```
$ snakemake  --use-conda --dryrun --runtime-profile inventory-runtime-local-master-june-15.txt
Building DAG of jobs...
Total time: 27.373994827270508 seconds
```
[inventory-runtime-local-master-june-15.txt](https://github.com/snakemake/snakemake/files/4781600/inventory-runtime-local-master-june-15.txt)

**Comment by johanneskoester** [comment]
Sure, sorry for that. So, I think we are ready for merging. This is very good overall.

**Comment by johanneskoester** [comment]
Thank you!

**Comment by vsoch** [comment]
Woohoo! Definitely agree.

**Pull request merged**